pyth-sdk/src/price.rs

```diff
@@ -897,4 +1201,966 @@ mod test {
+        );
+        succeeds(
+            pc(100 * (PD_SCALE as i64), 2 * PD_SCALE, -9),
+            100_000_000,
+            1_000_000_000_000_000_000,
+            100,
+            110,
+            -2,
+            pc(100 * (PD_SCALE as i64 - 10), 2 * PD_SCALE, -9),
+        );
+        succeeds(
+            pc(100 * (PD_SCALE as i64), 2 * PD_SCALE, -9),
+            1_000_000_000,
+            1_000_000_000_000_000_000,
+            100,
+            110,
+            -2,
+            pc(100 * (PD_SCALE as i64 - 10), 2 * PD_SCALE, -9),
+        );
+        // interpolation now doesn't lose precision, but normalize in final multiply loses precision
+        succeeds(
+            pc(100 * (PD_SCALE as i64), 2 * PD_SCALE, -9),
+            10_000_000_000,
+            1_000_000_000_000_000_000,
+            100,
+            110,
+            -2,
+            pc(100 * (PD_SCALE as i64), 2 * PD_SCALE, -9),
+        );
+        succeeds(
+            pc(100 * (PD_SCALE as i64), 2 * PD_SCALE, -9),
+            20_000_000_000,
+            1_000_000_000_000_000_000,
+            100,
+            110,
+            -2,
+            pc(100 * (PD_SCALE as i64), 2 * PD_SCALE, -9),
+        );
+        // precision no longer lost
+        succeeds(
+            pc(100 * (PD_SCALE as i64), 2 * PD_SCALE, -9),
+            100_000_000_000,
+            1_000_000_000_000_000_000,
+            100,
+            110,
+            -2,
+            pc(100 * (PD_SCALE as i64 + 10), 2 * PD_SCALE, -9),
+        );
+        succeeds(
+            pc(100 * (PD_SCALE as i64), 2 * PD_SCALE, -9),
+            200_000_000_000,
+            1_000_000_000_000_000_000,
+            100,
+            110,
+            -2,
+            pc(100 * (PD_SCALE as i64 + 20), 2 * PD_SCALE, -9),
+        );
+        succeeds(
+            pc(100 * (PD_SCALE as i64), 2 * PD_SCALE, -9),
+            1_000_000_000_000,
+            1_000_000_000_000_000_000,
+            100,
+            110,
+            -2,
+            pc(100 * (PD_SCALE as i64+100), 2 * PD_SCALE, -9),
+        );
+
+        // fails bc initial premium exceeds final premium
+        fails(
+            pc(100 * (PD_SCALE as i64), 2 * PD_SCALE, -9),
+            50,
+            100,
+            111,
+            110,
+            -2
+        );
+
+    }
+
+    #[test]
+    fn test_affine_combination() {
+        fn succeeds(x1: i64, y1: Price, x2: i64, y2: Price, x_query: i64, pre_add_expo: i32, expected: Price) {
+            let y_query = Price::affine_combination(x1, y1, x2, y2, x_query, pre_add_expo).unwrap();
+            
+            assert_eq!(y_query, expected);
+        }
+
+        fn fails(x1: i64, y1: Price, x2: i64, y2: Price, x_query: i64, pre_add_expo: i32) {
+            let result = Price::affine_combination(x1, y1, x2, y2, x_query, pre_add_expo);
+            assert_eq!(result, None);
+        }
+
+        // constant, inbounds
+        succeeds(
+            0,
+            pc(100, 0, -4),
```

Comment:
i think you want to test the confidences here too. (just make them price / 10 or something in all the tests)

Reply:
i see you tested confidence explicitly later, but I think you want to test the behavior of the confidence under all of the same conditions as the price itself, right?

Reply:
totally missed these comments--adding now